pkg/k8s/state_into_spec_annotation.go

```diff
@@ -19,24 +19,28 @@ import (
 	"strings"
 
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	"k8s.io/apimachinery/pkg/runtime/schema"
 )
 
-func ValidateOrDefaultStateIntoSpecAnnotation(obj *unstructured.Unstructured) error {
+// ValidateOrDefaultStateIntoSpecAnnotation validates the value of the
+// 'state-into-spec' annotation if it is set and defaults the annotation to the
+// passed in defaultValue if it is unset.
+func ValidateOrDefaultStateIntoSpecAnnotation(obj metav1.Object, gvk schema.GroupVersionKind, defaultValue string) error {
 	_, found := GetAnnotation(StateIntoSpecAnnotation, obj)
-	if !found {
-		SetAnnotation(StateIntoSpecAnnotation, StateMergeIntoSpec, obj)
+	if found {
+		return validateStateIntoSpecAnnotation(obj, gvk)
 	}
-	return validateStateIntoSpecAnnotation(obj, obj.GroupVersionKind())
+	defaultStateIntoSpecAnnotation(obj, gvk, defaultValue)
+	return nil
 }
 
-func EnsureSpecIntoSateAnnotation(obj *Resource) error {
-	_, found := GetAnnotation(StateIntoSpecAnnotation, obj)
-	if !found {
+func defaultStateIntoSpecAnnotation(obj metav1.Object, gvk schema.GroupVersionKind, defaultValue string) {
+	if defaultValue == StateAbsentInSpec && !ResourceSupportsStateAbsentInSpec(gvk.Kind) {
```

Comment:
I suggest we remove `ResourceSupportsStateAbsentInSpec` now that we support state-into-status.

IIRC the only reason this method exist is because we were not able to populate the ip address of ComputeAddress resource into status.

IMO we should update ComputeAddress CRD to surface that info in status so all resources will support state absent.

Reply:
Good point! Skipped the special handling here, and created b/322836859 and b/322839663 to remove the special handling completely.